### ISSUE TITLE
[chm-v1-preregister] EIEM optimized exact-retrieval matched-LM replication — seeds 19591/19592/19593

### ISSUE THREAD

**Issue by vinceackermann2-sys** [vinceackermann2-sys/tam-research#959]
## Classification and authority

`PREREGISTRATION_ONLY_NO_GPU_AUTHORITY`

Fresh scientific successor to completed systems issue #957 / merged PR #958. This protocol is independent of the stopped #854 sequence. Old scientific seeds 8611/8612/8613 remain permanently blocked and must never be reused.

This issue freezes the scientific question, fairness rules, evaluation, and stop conditions before any fresh scientific result is observed. It authorizes implementation, CPU smoke checks, and ordinary CI only. **It does not authorize GPU/Modal/paid compute.** A separate run-control issue with explicit owner authorization is required before any scientific GPU allocation.

Authoritative base at preregistration: `main` commit `2e1378f92048e2a3a32d0e36bf895c8846c6a20d` (merge of #958), exact-main CI run `34750818804`: 1389 passed / 15 warnings.

## Fresh scientific seeds

Seed family is deterministically issue-number-derived and will be frozen in the first protocol comment immediately after GitHub assigns this issue number, before implementation or scientific execution. The rule is:

`seed_k = 10000 + 10 * ISSUE_NUMBER + k`, for k in {1,2,3}.

Before freezing those values, audit GitHub issues/PRs/code for prior use. If any candidate is already used, do not substitute ad hoc; amend the derivation rule before implementation and before any scientific output.

A fresh seed is consumed once the scientific GPU function starts. Consumed seeds are never rerun or retried, including infrastructure/runtime failures after GPU allocation. Any replacement requires a separately recorded preregistration amendment made before replacement-seed scientific output exists.

## Scientific question

Replicate the original small matched-LM EIEM question under the same training/data/model/evaluation design, changing only the exact indexed inference implementation to the validated clipped-query kernel from #957/#958:

> Can a causal LM learn/use exact episodic memory outside its 512-token local attention window such that long-range state/rare-fact/multi-hop capability improves, while exact indexed inference reproduces flat-memory retrieval decisions, reads sparsely, and avoids the previously identified avoidable exact-index traversal overhead?

This is a **fresh replication**, not continuation of #854 and not a reinterpretation of seed 8611.

## Frozen model comparison

Per paired seed train two independently parameter-matched models:

1. **LOCAL** — bounded-local causal Transformer control, no episodic memory.
2. **EIEM-FLAT** — same bounded-local backbone plus learned normalized query/key address projections, exact episodic key/value store, exhaustive flat memory scoring, and learned memory integration.
3. **EIEM-OPT-INDEXED** — the exact same trained EIEM-FLAT checkpoint and episodic state; only exhaustive memory search is replaced by the exact clipped-query branch-and-bound implementation merged via #958.

EIEM-FLAT and EIEM-OPT-INDEXED must use identical learned addresses, distance function, tie-breaking, selected memory items, values, integration computation, weights, and state. Any discrete retrieval mismatch is a gate failure.

No oracle IDs, future information, generator labels, or hidden scoring metadata may enter the model/index at inference.

## Parameter/training fairness — unchanged from #854

Target approximately 25M trainable parameters.

Freeze:
- LOCAL and EIEM-FLAT trainable parameter counts within 1%;
- GPT-2 BPE vocabulary 50,257;
- same immutable FineWeb-Edu train/validation binaries already used by the repository;
- exactly **8,388,608 training tokens/model**;
- same paired seed controls initialization and deterministic batch stream;
- AdamW betas `(0.9, 0.95)`, weight decay `0.1`, peak LR `3e-4`, cosine decay, 2% warmup, grad clip 1.0;
- bfloat16 on GPU;
- identical global optimizer batch and token accounting across LOCAL/EIEM-FLAT;
- no tuning of architecture, optimizer, data, objective weights, memory capacity, retrieval-k, or thresholds after any fresh seed output exists.

## Long-memory protocol — unchanged

- dense local window: 512 tokens;
- held-out evaluation includes evidence beyond 512 tokens;
- episodic state reset at documented session boundaries;
- normalized learned query/key projections from causal model-visible representations;
- causal values available at write time;
- exact index contains only exact stored addresses/values plus metadata derivable from them;
- training may use differentiable flat memory scoring; optimized indexed retrieval is inference substitution only.

## Frozen evaluation families

Every completed seed reports:

### Ordinary language
- held-out FineWeb-Edu NLL/PPL on identical validation bytes;
- LOCAL vs EIEM-FLAT NLL delta;
- EIEM-FLAT vs EIEM-OPT-INDEXED logits/NLL equality within numerical tolerance when selected memories are identical.

### Long-range causal memory
Deterministic held-out generators/templates disjoint from training instances, required evidence outside the 512-token local window:
- rare fact;
- overwrite/stale state;
- two-hop;
- negative/no-memory-needed local control.

## Primary full-pass criteria — unchanged scientifically

A **FULL REPLICATION PASS** requires all of the following across the three fresh paired seeds:

1. **Exactness:** EIEM-OPT-INDEXED retrieved item(s) match EIEM-FLAT on 100% of evaluated retrieval calls for identical address states/queries.
2. **Long-range benefit:** EIEM-FLAT beats LOCAL by at least +10 percentage points absolute mean accuracy averaged over rare-fact, overwrite, and two-hop, and wins each family in at least 2/3 seeds.
3. **Stale-state:** EIEM-FLAT overwrite stale-value error at least 50% lower relative than LOCAL, unless LOCAL stale error <1%, in which case EIEM-FLAT may be no worse by >0.5 percentage points.
4. **Ordinary-language preservation:** mean EIEM-FLAT FineWeb-Edu NLL no worse than LOCAL by >+0.10, and no individual seed worse by >+0.15.
5. **Local negative control:** EIEM-FLAT local-only accuracy no worse than LOCAL by >2 percentage points on any seed.
6. **Sparse reads:** at largest preregistered memory size, EIEM-OPT-INDEXED exact address-vector reads <=25% of flat exhaustive reads on mean; metadata traversal reported separately.
7. **Systems sanity:** no NaN/Inf, cross-session aliasing, hidden persistent state, or base-parameter mutation during inference updates.

## Systems measurement specific to this replication

Because #854 stopped on practical indexed overhead, every fresh seed must also report end-to-end flat vs optimized-indexed inference wall-clock at batch 1 and the same throughput-oriented batch, including:
- index build/update;
- write;
- search;
- wrapper/integration overhead;
- exact address reads;
- directory nodes visited.

The #957 microbenchmark success is prior systems evidence only; it is **not** substituted for end-to-end scientific-run timing.

## Sequential execution / stop rules

Run seeds strictly one at a time. Seed 1 must complete and be terminalized before seed 2 is authorized; seed 2 before seed 3.

Stop before later seeds if any completed seed shows:
- indexed/flat discrete retrieval mismatch;
- oracle/future/generator metadata leakage;
- numerical instability or fairness violation;
- ordinary-language degradation beyond +0.15 NLL for that seed;
- local negative-control degradation >2 percentage points;
- optimized indexed read fraction >25% at the largest memory size;
- optimized indexed end-to-end practical overhead still erases any usable inference advantage relative to the corresponding flat path;
- any implementation defect that changes the frozen scientific comparison.

Additionally, after two completed seeds, stop if any 2/3 family-win criterion has become mathematically impossible to satisfy on seed 3.

No automatic retries or redispatches. Infrastructure failure after GPU allocation consumes that seed and is not scientific evidence.

## Mandatory measurements

Retain per seed/model/mode:
- trainable parameter count;
- exact train tokens and validation data identity;
- train loss/NLL trajectory;
- held-out NLL/PPL;
- rare-fact/overwrite/stale/two-hop/local-control metrics by distance and memory size;
- flat/indexed retrieval match rate;
- address vectors scored/query;
- directory nodes/query;
- build/update/write/search/wrapper timing;
- batch-1 and throughput inference timing;
- training wall-clock, tokens/s, peak VRAM, compile time if applicable;
- memory-state bytes/session;
- all failed/unfavorable attempts and consumed seeds.

## Execution discipline

1. Freeze fresh seeds in this issue before implementation.
2. Implement on a fresh branch from authoritative `main`.
3. Add CPU-only smoke/correctness tests and preserve the scientific runner's no-retry seed-consumption marker discipline.
4. Merge only after green CI on exact implementation head and verify exact-main CI.
5. Create a separate run-control issue naming the merged commit, first fresh seed, GPU class, maximum budget, and explicit owner authorization.
6. Only then may the first scientific GPU function start.

## Interpretation ceiling

Even FULL REPLICATION PASS means only that this ~25M matched-LM setup shows reproducible long-memory signal plus exact sparse optimized inference under the frozen protocol. It does not establish novelty, SOTA, large-scale training efficiency, production speedup, AGI, RSI, or an architecture breakthrough.

**Comment by vinceackermann2-sys** [comment]
## Fresh scientific seed freeze

Issue #959's deterministic derivation rule gives the fresh paired scientific seeds:
- seed 1: `19591`
- seed 2: `19592`
- seed 3: `19593`

Freshness audit completed **before implementation and before any fresh scientific output**:
- GitHub issue searches: no prior use of these seed values found;
- GitHub PR search: no prior use found;
- GitHub commit searches: no prior use found;
- GitHub code search is currently returning an upstream `502`, so it is unavailable and is **not** treated as either positive or negative freshness evidence, matching the conservative audit practice used for #854.

These values were derived only after GitHub assigned issue number `959`. They are now frozen and reserved exclusively for this #959 replication. They are not continuations or replacements of `8611/8612/8613`; those old seeds remain permanently blocked under the stopped #854 sequence.

A #959 seed becomes consumed only when its scientific GPU function actually starts. At this freeze point **19591, 19592, and 19593 are unconsumed**. No GPU/Modal/paid scientific execution is authorized by this freeze comment.

**Comment by vinceackermann2-sys** [comment]
CHM_V1_959_IMPLEMENTATION_AUTHORITY_V1

classification=IMPLEMENTATION_COMPLETE_NO_GPU_AUTHORITY
pr=960
implementation_head=fce0c3770cb943938852531961c55f6b16f7f0e7
merge_commit=22da370b335727973e76fd20d0a3f87f6bb2d320
merge_tree=a217d3f88e6784b491c7f3dc4333a22283c2583f
pr_ci_run=34766809641
pr_ci_job=103749119925
pr_ci_conclusion=success
exact_main_ci_run=34767186601
exact_main_ci_job=103750143188
exact_main_ci_conclusion=success
changed_files=2_additive_only
frozen_scientific_model_training_core_changed=false
fresh_scientific_seeds=19591,19592,19593
fresh_seed_19591_consumed=false
fresh_seed_19592_consumed=false
fresh_seed_19593_consumed=false
legacy_seeds_8611_8612_8613_authorized=false
gpu_allocated=false
modal_or_paid_execution=false

The #959 CPU-only optimized exact-retrieval integration is now merged and verified on exact current main. The implementation adds only `tam_research/chm_v1_optimized_replication.py` and `tests/test_chm_v1_optimized_replication.py`; the frozen #854 model/training implementation is unchanged. No scientific seed has been executed or consumed.

Training remains BLOCKED. Per the frozen #959 protocol, the next execution step requires a separate run-control issue that names this merged commit, seed 19591, an explicit GPU class and maximum budget, and a new explicit owner authorization for paid/GPU scientific execution. A generic request to continue implementation is not recorded here as paid/GPU authorization.

**Comment by vinceackermann2-sys** [comment]
CHM_V1_959_PROPOSED_SEED_19591_RUN_ENVELOPE_NO_AUTHORITY

classification=PROPOSED_RUN_CONTROL_ENVELOPE_ONLY
scientific_seed=19591
scientific_seed_consumed=false
run_control_issue_created=false
paid_compute_authorized=false
gpu_allocated=false
source_main_sha=22da370b335727973e76fd20d0a3f87f6bb2d320
source_main_ci=34767186601
pr=960

Proposed execution envelope, intentionally NOT authorized by this comment:
- provider: Modal
- GPU: 1x NVIDIA L4
- CPU reservation: 4 physical cores
- memory reservation: 8 GiB
- max GPU/runtime allocation: 3,600 seconds for seed 19591
- max billed-compute cap: USD 1.25 for seed 19591
- sequential seed discipline: seed 19591 only; 19592/19593 remain blocked until 19591 is terminalized and #959 stop rules are applied
- no automatic retry / rerun / redispatch
- seed becomes consumed at first scientific GPU-function start / GPU allocation, per #959
- old seeds 8611/8612/8613 remain permanently blocked

Rationale:
- L4 matches the historical #854/#903 GPU class, improving systems comparability.
- Current public Modal rates remain approximately L4 $0.000222/s, CPU $0.0000131/core/s, memory $0.00000222/GiB/s. At 4 cores + 8 GiB + one L4 for 3,600 s, the conservative resource-cost arithmetic is about $1.051776 before any plan credits; $1.25 is the proposed hard budget envelope.

AUTHORIZATION BOUNDARY
This comment is preparation only. It MUST NOT be treated as owner authorization. Do not create an authorized run-control issue, implement/merge a paid launcher, dispatch Modal, allocate an L4, or consume seed 19591 until the user explicitly authorizes this exact or a revised paid/GPU envelope in chat. A bare “continue” is not sufficient paid-compute authorization.

**Comment by vinceackermann2-sys** [comment]
CHM_V1_959_EXPLICIT_OWNER_AUTHORIZATION_SEED_19591

classification=AUTHORIZED_PAID_GPU_EXECUTION_ENVELOPE
source_main_sha=22da370b335727973e76fd20d0a3f87f6bb2d320
source_main_ci=34767186601
scientific_seed=19591
provider=Modal
gpu=1x NVIDIA L4
cpu=4 physical cores
memory=8 GiB
max_gpu_runtime_seconds=3600
max_billed_compute_usd=1.25
automatic_retry=false
seed_19592_authorized=false
seed_19593_authorized=false
legacy_8611_8612_8613_authorized=false

Explicit owner authorization was received in chat verbatim: “I authorize seed 19591 on 1× L4, max 3,600 seconds and max $1.25 paid compute. Proceed.”

This authorization permits creation of the separate run-control issue, implementation/review/merge of a guarded one-shot launcher/workflow under fresh CI, and—only after exact-current-main source/authorization/corpus/result-namespace/seed-consumption preflight passes—one scientific GPU start for seed 19591 within the envelope above.

Seed 19591 remains unconsumed at this authorization-record step. It becomes consumed when the scientific GPU function starts / GPU allocation begins. No retry/rerun/redispatch is authorized after consumption. Seeds 19592 and 19593 remain blocked until 19591 is terminalized and #959 stop rules are applied.

**Comment by vinceackermann2-sys** [comment]
CHM_V1_959_OWNER_AUTHORIZATION_RECEIVED_SEED_19591

Explicit owner authorization received in chat on 2026-09-13:

> “I authorize seed 19591 on 1× L4, max 3,600 seconds and max $1.25 paid compute. Proceed.”

Authorized envelope:
- seed: 19591 only
- provider: Modal
- GPU: 1× NVIDIA L4
- CPU reservation: 4 physical cores
- memory reservation: 8 GiB
- max runtime: 3,600 seconds
- max billed compute: USD 1.25
- no automatic retry / rerun / redispatch
- 19592/19593 remain blocked
- 8611/8612/8613 remain blocked

This authorizes creation of the separate run-control issue and guarded launcher/workflow implementation. Seed 19591 remains unconsumed until the scientific GPU function actually starts / GPU allocation begins.

**Comment by vinceackermann2-sys** [comment]
CHM_V1_959_TERMINAL_SEED_19591_RESULT_V1

classification=SEED_19591_STOP_BEFORE_19592
workflow_run=34889594634
workflow_conclusion=success
source_sha=c846e2fb0ec774cc1c95ccde2979dbde08d254b1
source_tree=31233258194e5729c7d24e4c24460b2b000454ac
harness_blob=28e8bf8afaec1bcb18b8794f0a15ebfafb8cb898
run_control_issue=966
trigger_issue=968
result_comment=5670294388
scientific_seed_19591_consumed=true
scientific_seed_19592_consumed=false
scientific_seed_19593_consumed=false
scientific_seed_19592_authorized=false
scientific_seed_19593_authorized=false
automatic_retry=false

Authoritative seed-19591 result under the frozen #959 protocol:
- LOCAL NLL: 6.338648058474064
- EIEM-FLAT NLL: 6.334924548864365 (delta vs LOCAL -0.0037235096096992493)
- EIEM-OPT-INDEXED NLL: 6.334924548864365 (delta vs flat 0.0)
- optimized/flat discrete retrieval exact match: 1.0
- sparse address-read fraction at 1024 items: 0.16107671453692848
- flat practical timing: 20.921868700999994 s
- optimized-indexed practical timing: 51.258811986999945 s (~2.4500x flat; ~145.0% slower)
- rare-fact accuracy LOCAL/EIEM: 0.125 / 0.125
- overwrite accuracy LOCAL/EIEM: 0.125 / 0.125
- two-hop accuracy LOCAL/EIEM: 0.125 / 0.125
- local-negative accuracy LOCAL/EIEM: 0.125 / 0.125
- systems sanity gates: PASS (no NaN/Inf, no cross-session aliasing, no hidden persistent state, no inference base-parameter mutation)

Preregistered stop rule fired: `index_overhead_erases_practical_advantage`. The exact optimized index did preserve retrieval decisions and reduced address-vector reads below the <=25% gate, but its end-to-end practical cost was materially worse than the corresponding flat path. Therefore the #959 sequence terminates after seed 19591. Do not run seed 19592 or 19593 under this protocol; do not retry/rerun/redispatch seed 19591.

Interpretation ceiling: this is one completed ~25M matched-LM replication seed and a preregistered STOP result. It is not a full replication pass, scaling result, novelty/SOTA result, AGI/RSI evidence, or breakthrough claim.
